---
title: ""
execute:
  echo: false
jupyter: python3
output: word_document
cap-location: bottom
format: 
  html:
    page-layout: full
  pdf:
    pdf-engine: lualatex
    fontsize: 12pt
    mainfont: Open Sans
    sansfont: Open Sans
    # classoption: landscape
    pagestyle: empty
    include-in-header: 
      text: |
        \usepackage{lscape}
        \newcommand{\blandscape}{\begin{landscape}}
        \newcommand{\elandscape}{\end{landscape}}
  
---


```{python}
#| label: tbl-hlahd_onlymatched_threashold_2F
#| tbl-cap: This table describe the number of calls obtained from HLA-HD (**results DOWNSCALED to 2 fields**) regarding the quality of alleles. Threshold usually used by the community are the sum of the number of reads for allele 1 & 2 (minimum of 20) and individual number of reads for allele 1 or 2 (minimum of 10). 
#| 
import pandas
import numpy as np
from IPython.display import Markdown, display
def printmd(string):
    display(Markdown(string))

def hlahd_onlymatched_threashold_2F():
  rez_hlahd = pandas.read_csv('data/OUTPUT/5.N_1_RESULTS/all_results-2F.csv')
  genes_list = sorted(rez_hlahd.gene.unique().tolist())
  markdown_text="| Gene | total calls | calls < threshold | calls < threshold (%) |\n"
  markdown_text+="| --- | --- | --- | --- |\n"

  for n,gene in enumerate(genes_list ):
    # if gene in ["DRB2","DRB3","DRB4","DRB5"]:
    #   continue
    hlahd_results = rez_hlahd.query("gene == @gene")
    hlahd_results = hlahd_results.loc[(hlahd_results['comparison_int'] == 7)]

    ## nb available calls
    nb_calls_all_hlahd_results = hlahd_results.shape[0]
    # if nb_calls_all_hlahd_results == 0:
    #   continue
    ## only calls > threshold_1
    hlahd_good_results = hlahd_results.loc[(hlahd_results['hlahd_nbreads1'] >= 10) & (hlahd_results['hlahd_sum_reads'] >= 20 )].shape[0]
    ## percentage calls < threshold_1
    percentage_calls_bad_hlahd_results = (nb_calls_all_hlahd_results - hlahd_good_results)/nb_calls_all_hlahd_results*100 if nb_calls_all_hlahd_results > 0 else 0
    # ## number of unique alleles
    # col_split = hlahd_good_results.hlahd_alleles.str.split('_', expand=True)
    # nb_unique_alleles = pandas.concat([col_split[0], col_split[1]], axis=0).nunique() if nb_calls_all_hlahd_results > 0 else 0
    # ## percentage of homozigosity
    # percentage_calls_hmz =  hlahd_good_results.loc[hlahd_good_results['hlahd_genotype'] == 'Hmz'].shape[0]/nb_calls_all_hlahd_results*100 if nb_calls_all_hlahd_results > 0 else 0
    
    markdown_text+=f"|{gene}|{nb_calls_all_hlahd_results}|{(nb_calls_all_hlahd_results - hlahd_good_results)}|{percentage_calls_bad_hlahd_results:.2f}|\n"
    
  printmd(markdown_text)
  # display(df_hlahd)
hlahd_onlymatched_threashold_2F()
```